src/dmd/mars.d

```diff
@@ -2125,3 +2168,278 @@ private bool parseCommandLine(const ref Strings arguments, const size_t argc, re
+// array of module patterns used to include/exclude imported modules
+private __gshared Array!(const(char)*) includeModulePatterns;
+private __gshared Modules compiledImports;
+private extern(C++) bool marsOnImport(Module m)
+{
+    if (includeImports)
+    {
+        Identifiers empty;
+        if (includeImportedModuleCheck(ModuleComponentRange(
+            (m.md && m.md.packages) ? m.md.packages : &empty, m.ident, m.isPackageFile)))
+        {
+            if (global.params.verbose)
+                fprintf(global.stdmsg, "compileimport (%s)\n", m.srcfile.toChars);
+            compiledImports.push(m);
+            return true; // this import will be compiled
+        }
+    }
+    return false; // this import will not be compiled
+}
+
+// A range of component identifiers for a module
+private struct ModuleComponentRange
+{
+    Identifiers* packages;
+    Identifier name;
+    bool isPackageFile;
+    size_t index;
+    @property auto totalLength() const { return packages.dim + 1 + (isPackageFile ? 1 : 0); }
+
+    @property auto empty() { return index >= totalLength(); }
+    @property auto front() const
+    {
+        if (index < packages.dim)
+            return (*packages)[index];
+        if (index == packages.dim)
+            return name;
+        else
+            return Identifier.idPool("package");
+    }
+    void popFront() { index++; }
+}
+
+/*
+ * Determines if the given module should be included in the compilation.
+ * Returns:
+ *  True if the given module should be included in the compilation.
+ */
+private bool includeImportedModuleCheck(ModuleComponentRange components)
+    in { assert(includeImports); } body
+{
+    createMatchNodes();
+    size_t nodeIndex = 0;
+    while (nodeIndex < matchNodes.dim)
+    {
+        //printf("matcher ");printMatcher(nodeIndex);printf("\n");
+        auto info = matchNodes[nodeIndex++];
+        if (info.depth <= components.totalLength())
+        {
+            size_t nodeOffset = 0;
+            for (auto range = components;;range.popFront())
+            {
+                if (range.empty || nodeOffset >= info.depth)
+                {
+                    // MATCH
+                    //printf("matcher ");printMatcher(nodeIndex - 1);
+                    //printf(" MATCHES module '");components.print();printf("'\n");
+                    return !info.isExclude;
+                }
+                if (!range.front.equals(matchNodes[nodeIndex + nodeOffset].id))
+                {
+                    break;
+                }
+                nodeOffset++;
+            }
+        }
+        //printf("matcher ");printMatcher(nodeIndex-1);
+        //printf(" does not match module '");components.print();printf("'\n");
+        nodeIndex += info.depth;
+    }
+    assert(nodeIndex == matchNodes.dim, "code bug");
+    return includeByDefault;
+}
+
+// Matching module names is done with an array of matcher nodes.
+// The nodes are sorted by "component depth" from largest to smallest
```

Comment:
What is a "component depth"?

Reply:
The "depth" is the number of components in a module name (i.e. `foo.bar.baz` has 3 components, so a depth of 3).  It's been so long I'm not sure why the term "depth" was chosen.